.circleci/config.yml

```diff
@@ -1,5 +1,106 @@
+    environment:
+      - TEST_RESULTS: /tmp/test-results  # path to where test results are saved
+
+# reusable 'commands' to be added as a step in jobs
+commands:
+  integration-test:
+    description: Run integration tests
+    parameters:
+      version:
+        description: what branch/version/tag to checkout
+        type: string
+        default: master
+      checkout-url:
+        description: HTTPS checkout URL (SSH checkout if using SSH key)
+        type: string
+      checkout-path:
+        description: path to checkout code
+        type: string
+      test-command:
+        description: command to run tests
+        type: string
+    steps:
+        - run: git clone << parameters.checkout-url >> --branch << parameters.version >> << parameters.checkout-path >>
+        
+        # get consul binary
+        - attach_workspace:
+            at: /go/bin
+        # make dev build of nomad
+        - run:
+            command: make dev
+            working_directory: << parameters.checkout-path >>
+
+        - run:
+            name: install junit and setup test path
+            command: |
+              go get github.com/jstemmer/go-junit-report
+              mkdir -p /tmp/test-results
+        
+        # run integration tests
+        - run: 
+            command: << parameters.test-command >> | tee ${TEST_RESULTS}/results.out
+            working_directory: << parameters.checkout-path >>
+
+
+        # generate XML results
+        - run:
+            command: go-junit-report < ${TEST_RESULTS}/results.out > ${TEST_RESULTS}/results.xml
+            when: always
+        
+        # store test results for CircleCI
+        - store_test_results:
+            path: /tmp/test-results
+
+
 jobs:
+  dev-build:
+    executor: go
+    working_directory: /go/src/github.com/hashicorp/consul
+    steps:
+      - checkout
+      - run: make dev
+
+      # save dev build to CircleCI
+      - store_artifacts:
+          path: /go/bin/consul
+
+      # save dev build to pass to downstream jobs
+      - persist_to_workspace:
+          root: /go/bin
+          paths:
+            - consul
+
+  # Nomad 0.8 builds on go0.10
+  # Run integration tests on nomad/v0.8.7
+  nomad-integration-0_8:
+      docker:
+        - image: circleci/golang:1.10
+      environment:
+        - TEST_RESULTS: /tmp/test-results
+      steps:
+        - integration-test:
+            version: v0.8.7
+            checkout-url: https://github.com/hashicorp/nomad.git
+            checkout-path: /go/src/github.com/hashicorp/nomad
+            test-command: go test -v ./command/agent/consul -run TestConsul
+
+  # run integration tests on nomad/master
+  nomad-integration-master:
+      executor: go
+      steps:
+        - integration-test:
+            version: master
+            checkout-url: https://github.com/hashicorp/nomad.git
+            checkout-path: /go/src/github.com/hashicorp/nomad
+            test-command: go test -v ./command/agent/consul -run TestConsul
```

Comment:
👋 - Thanks for getting started! but there are a couple of issues with this.

The `command/agent/consul` tests do not actually start a copy of consul, so these won't be effected by changing the consul binary in the `PATH`. They [start][1] a fake consul [implementation][2].

The only Nomad `unit` tests that spin up a real consul agent are the ones used to test [Consul Template][3]. However, you should be able to run the `e2e` `consul` and `consultemplate` tests without having to spin up a remote cluster by starting consul and nomad dev agents, then running `go test -v ./e2e -run TestE2E/Consul` and `go test -v ./e2e -run TestE2E/Consul_Template`.
The caveat here, is that you'd need to build the nomad binary for $NOMAD_VERSION, but then run the tests from `master`, due to the lack of e2e tests in prior nomad revisions.

[1]: https://github.com/hashicorp/nomad/blob/3477ea009f81f42d2ba506e50f352d9ccacf7435/command/agent/consul/unit_test.go#L116-L127
[2]: https://github.com/hashicorp/nomad/blob/3477ea009f81f42d2ba506e50f352d9ccacf7435/command/agent/consul/catalog_testing.go#L53
[3]: https://github.com/hashicorp/nomad/blob/81e259810974a7f9dfca16defba0206d1d807dff/client/allocrunner/taskrunner/template/template_test.go#L141

Reply:
Does it? When @pearkes and I were running the `command/agent/consul` tests we could see a consul process start up which presumably used the `consul` in your PATH (In this case, the one `make dev` creates and sticks in `/go/bin`).

The steps you outlined to get the integration seems pretty simple. I was under the impression it may be more work than that so I can tweak this to run that if we decide the agent tests don't do enough.

Reply:
@dantoml Great to have your feedback here, thanks! If I understand correctly there are some Nomad / Consul integration tests that are mocked and some that are not. In any test that uses the Consul testutil server (https://github.com/hashicorp/nomad/blob/3a491ab2b4924dc16f523512949a47b6b0a07254/command/agent/consul/int_test.go#L45-L51) you'd be using whatever Consul binary is on the path.

I ran through this locally a few days ago and it shows the bug on master and then being fixed on our branch.

https://gist.github.com/pearkes/8bd9d2b37338879a706906abfe8f9f90

That is testing the same package as Circle does here.

It feels like this is a bit of a stand-in until we can collectively rely on the e2e tests for pass/fail on master -- which seems a bit out of scope for Consul PRs. I kind of like the short-term value of this for now, what do you think? Also let me know if I'm missing something subtle here with how Nomad is invoking Consul -- but it feels like that test should have passed with both Consul binaries it if it was fully mocked.

Reply:
@dantoml would you be OK with getting this in and I can work with you to either replace or add the E2E tests on to this?

Reply:
Talked with @dantoml offline and I am ok to merge this. We will work on getting the E2E tests in at some point in the future.